File: igct/trie/impl.py
__author__ = 'mactep'
from collections.abc import Sized

from igct.trie.node import Node


class TrieImpl(Sized):
    def __init__(self):
        self._root = Node(None, None, 0)
        self._size = 1

    def insert(self, node, symbol):
        if symbol not in node:
            node[symbol] = Node(node, symbol, self._size)
            self._size += 1
        return node[symbol]

    @property
    def root(self):
        return self._root

    def __len__(self):
        return self._size

File: scalatest/src/Impl.scala
/**
 * Created with IntelliJ IDEA.
 * User: mactep
 * Date: 04.10.13
 * Time: 12:05
 */
class Impl {
  private val _root : Node = new Node()
  private var _size : Int  = 1

  def insert(node : Node, symbol : Byte) : Node = {
    if (!node.contains(symbol)) {
      node.set(new Node(node, symbol, _size))
      _size += 1
    }
    return node.get(symbol)
  }

  def root() : Node = {
    return _root
  }

  def size() : Int = {
    return _size
  }
}
